File: rmd-examples/java-example/src/main/java/example/Application.java
package example;

import java.util.concurrent.ExecutionException;

import static net.uoit.rmd.Rmd.*;

public class Application {

    public static void main(String[] args) throws InterruptedException, ExecutionException {
        System.out.println("6! " + delegate(Functions::factorial, 6));

        for (int i = 0; i < 10; i++) {
            int finalI = i;
            delegate((a) -> a * 2, i, n -> System.out.println(finalI + " * 2 = " + n));
        }

        waitForAsyncJobs();
        System.exit(0);
    }
}

File: rmd-examples/kotlin-example/src/main/kotlin/example/Application.kt
package example

import net.uoit.rmd.async
import net.uoit.rmd.callback
import net.uoit.rmd.waitForAsyncJobs
import java.util.*
import kotlin.collections.ArrayList


fun main(args: Array<String>) {
    val numbers = 10.randomList(0, 100)
    println("List of Numbers: $numbers")

    sortJob(numbers)

    waitForAsyncJobs()
    System.exit(0)
}

fun sortJob(list: MutableList<Int>) = async {
    list.sort()
    list
} callback {
    println("Sort Result: $it")
}

fun IntRange.random() = Random().nextInt((endInclusive + 1) - start) + start

fun Int.randomList(min: Int, max: Int): MutableList<Int> {
    val numbers: ArrayList<Int> = ArrayList()

    for (i in 0..this) {
        numbers.add((min..max).random())
    }

    return numbers
}
